{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'2765658': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGATGAGACCGTGG...CGC', SingleLetterAlphabet()), '2765657': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAG...GGC', SingleLetterAlphabet()), '2765656': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TAA', SingleLetterAlphabet()), '2765655': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAAACAACAT...CAT', SingleLetterAlphabet()), '2765654': Seq('ACGGCGAGCTGCCGAAGGACATTGTTGAGACAGCAGAATATACGATTGAGTGAA...AAA', SingleLetterAlphabet()), '2765652': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...CCC', SingleLetterAlphabet()), '2765651': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...TGT', SingleLetterAlphabet()), '2765650': Seq('TGTTGAGATAGCAGAATATACATCGAGTGAATCCGGAGGACCTGTGGTTATTCG...GCA', SingleLetterAlphabet()), '2765649': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATAGTAG...AGC', SingleLetterAlphabet()), '2765648': Seq('CGTAACCAGGTTTCCGTAGGTGAACCTGCGGCAGGATCATTGTTGAGACAGCAG...AAG', SingleLetterAlphabet()), '2765647': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...GAG', SingleLetterAlphabet()), '2765646': Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAGAATATATGATCGAGT...ACC', SingleLetterAlphabet()), '2765645': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGCAG...TTT', SingleLetterAlphabet()), '2765644': Seq('ATATGATCGAGTGAATCTGGTGGACTTGTGGTTACTCAGCTCGCCATAGGCTTT...TTA', SingleLetterAlphabet()), '2765643': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATAGTAG...TCC', SingleLetterAlphabet()), '2765642': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAG...AGC', SingleLetterAlphabet()), '2765641': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAGTAT...TAA', SingleLetterAlphabet()), '2765640': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGCTGAGACCGTAG...AGC', SingleLetterAlphabet()), '2765639': Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...CTA', SingleLetterAlphabet()), '2765638': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GAG', SingleLetterAlphabet()), '2765637': Seq('CGTAACAAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAAGCCCCCA...AGC', SingleLetterAlphabet()), '2765636': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCCCCA...GGA', SingleLetterAlphabet()), '2765635': Seq('CTAACCAGGGTTCCGAGGTGACCTTCGGGAGGATTCCTTTTTAAGCCCCCGAAA...TTA', SingleLetterAlphabet()), '2765634': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...GGA', SingleLetterAlphabet()), '2765633': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TGA', SingleLetterAlphabet()), '2765632': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCCCCA...TGA', SingleLetterAlphabet()), '2765631': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...TGA', SingleLetterAlphabet()), '2765630': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCCA...TTT', SingleLetterAlphabet()), '2765629': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTTCGGAAGGATCATTGTTGAGACCGCAA...TAA', SingleLetterAlphabet()), '2765628': Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCCTTGTTGAGACCGCCA...TAA', SingleLetterAlphabet()), '2765627': Seq('CGTAACCAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGACCGCCA...CGC', SingleLetterAlphabet()), '2765626': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACCGCAA...AGA', SingleLetterAlphabet()), '2765625': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGCTCATTGTTGAGACCGCAA...AAG', SingleLetterAlphabet()), '2765624': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAGGGATCATTGTTGAGATCGCAT...ACC', SingleLetterAlphabet()), '2765623': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAAGGTCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765622': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765621': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC', SingleLetterAlphabet()), '2765620': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GTG', SingleLetterAlphabet()), '2765619': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGGTCGCAT...AAG', SingleLetterAlphabet()), '2765618': Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...CCC', SingleLetterAlphabet()), '2765617': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...ATA', SingleLetterAlphabet()), '2765616': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...AGC', SingleLetterAlphabet()), '2765615': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA', SingleLetterAlphabet()), '2765614': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGC', SingleLetterAlphabet()), '2765613': Seq('CTGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGCAATAATTGATCGA...GCT', SingleLetterAlphabet()), '2765612': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TAA', SingleLetterAlphabet()), '2765611': Seq('CGTCACGAGGTTTCCGTAGGTGAATCTGCGGGAGGATCATTGTTGAGATCACAT...TGA', SingleLetterAlphabet()), '2765610': Seq('CTGAACCTGGTGTCCGAAGGTGAATCTGCGGATGGATCATTGTTGAGATATCAT...GTA', SingleLetterAlphabet()), '2765609': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGGGGAAGGATCATTGTTGAGATCACAT...TTT', SingleLetterAlphabet()), '2765608': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA', SingleLetterAlphabet()), '2765607': Seq('TCTACTGCAGTGACCGAGATTTGCCATCGAGCCTCCTGGGAGCTTTCTTGCTGG...GCA', SingleLetterAlphabet()), '2765606': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA', SingleLetterAlphabet()), '2765605': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA', SingleLetterAlphabet()), '2765604': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGT', SingleLetterAlphabet()), '2765603': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCAGTGTTGAGATCACAT...GGC', SingleLetterAlphabet()), '2765602': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC', SingleLetterAlphabet()), '2765601': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC', SingleLetterAlphabet()), '2765600': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GGT', SingleLetterAlphabet()), '2765599': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACGT...CTT', SingleLetterAlphabet()), '2765598': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG', SingleLetterAlphabet()), '2765597': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765596': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765595': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT', SingleLetterAlphabet()), '2765594': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GTT', SingleLetterAlphabet()), '2765593': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCGCAT...GTT', SingleLetterAlphabet()), '2765592': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGA', SingleLetterAlphabet()), '2765591': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...CCC', SingleLetterAlphabet()), '2765590': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC', SingleLetterAlphabet()), '2765589': Seq('CGTAACAAGGTTTCCGTAGGTGAGCGGAAGGGTCATTGTTGAGATCACATAATA...AGC', SingleLetterAlphabet()), '2765588': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGTTCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765587': Seq('CGTCACGAGGTCTCCGGATGTGACCCTGCGGAAGGATCATTGTTGAGATCACAT...CAT', SingleLetterAlphabet()), '2765586': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TAA', SingleLetterAlphabet()), '2765585': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...TTA', SingleLetterAlphabet()), '2765584': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTT', SingleLetterAlphabet()), '2765583': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TTG', SingleLetterAlphabet()), '2765582': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...GAG', SingleLetterAlphabet()), '2765581': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765580': Seq('CGTAACCAGGTTTCCGTAGGTGGACCTTCGGGAGGATCATTTTTGAGATCACAT...GCA', SingleLetterAlphabet()), '2765579': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AAC', SingleLetterAlphabet()), '2765578': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA', SingleLetterAlphabet()), '2765577': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...GCA', SingleLetterAlphabet()), '2765576': Seq('CGTAACAAGGTTTCCGTAGGTGTACCTCCGGAAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765575': Seq('GGAAGGATCATTGCTGATATCACATAATAATTGATCGAGTTAAGCTGGAGGATC...GAG', SingleLetterAlphabet()), '2765574': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGC', SingleLetterAlphabet()), '2765573': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG', SingleLetterAlphabet()), '2765572': Seq('CGTAACAAGGATTCCGTAGGTGAACCTGCGGGAGGATCATTGTTGAGATCACAT...AGC', SingleLetterAlphabet()), '2765571': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTCCGGAAGGATCATTGTTGAGATCACAT...CCC', SingleLetterAlphabet()), '2765570': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...TGT', SingleLetterAlphabet()), '2765569': Seq('CGTAACAAGGTTTCCGTAGGGTGAACTGCGGAAGGATCATTGTTGAGATCACAT...ATT', SingleLetterAlphabet()), '2765568': Seq('CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACAT...AGG', SingleLetterAlphabet()), '2765567': Seq('GTAGGTGAACCTGCGGAAGGATCATTGTTGAGATCACATAATAATTGATCGAGT...AGT', SingleLetterAlphabet()), '2765566': Seq('GGAAGGTCATTGCCGATATCACATAATAATTGATCGAGTTAATCTGGAGGATCT...GAG', SingleLetterAlphabet()), '2765565': Seq('CGTAACAAGGTTTCCGTAGGTGGACCTCCGGGAGGATCATTGTTGAGATCACAT...GCA', SingleLetterAlphabet()), '2765564': Seq('CATTGTTGAGATCACATAATAATTGATCGAGTTAATCTGGAGGATCTGTTTACT...GCC', SingleLetterAlphabet())}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'\\nIf you were to call the FetchGenbankFile with the following id:\\n\\nx=FetchGenbankFile(\\'186972394\\')\\n\\nYou should get a file named 186972394.gbk.txt in your computer with the Genbank file below.\\n\\nFrom the tutorial:\\n\\n>>> from Bio import Entrez\\n>>> Entrez.email = \"A.N.Other@example.com\"     # Always tell NCBI who you are\\n>>> handle = Entrez.efetch(db=\"nucleotide\", id=\"186972394\", rettype=\"gb\", retmode=\"text\")\\n>>> print handle.read()\\nLOCUS       EU490707                1302 bp    DNA     linear   PLN 05-MAY-2008\\nDEFINITION  Selenipedium aequinoctiale maturase K (matK) gene, partial cds;\\n            chloroplast.\\nACCESSION   EU490707\\nVERSION     EU490707.1  GI:186972394\\nKEYWORDS    .\\nSOURCE      chloroplast Selenipedium aequinoctiale\\n  ORGANISM  Selenipedium aequinoctiale\\n            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;\\n            Spermatophyta; Magnoliophyta; Liliopsida; Asparagales; Orchidaceae;\\n            Cypripedioideae; Selenipedium.\\nREFERENCE   1  (bases 1 to 1302)\\n  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,\\n            Endara,C.L., Williams,N.H. and Moore,M.J.\\n  TITLE     Phylogenetic utility of ycf1 in orchids\\n  JOURNAL   Unpublished\\nREFERENCE   2  (bases 1 to 1302)\\n  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,\\n            Endara,C.L., Williams,N.H. and Moore,M.J.\\n  TITLE     Direct Submission\\n  JOURNAL   Submitted (14-FEB-2008) Department of Botany, University of\\n            Florida, 220 Bartram Hall, Gainesville, FL 32611-8526, USA\\nFEATURES             Location/Qualifiers\\n     source          1..1302\\n                     /organism=\"Selenipedium aequinoctiale\"\\n                     /organelle=\"plastid:chloroplast\"\\n                     /mol_type=\"genomic DNA\"\\n                     /specimen_voucher=\"FLAS:Blanco 2475\"\\n                     /db_xref=\"taxon:256374\"\\n     gene            <1..>1302\\n                     /gene=\"matK\"\\n     CDS             <1..>1302\\n                     /gene=\"matK\"\\n                     /codon_start=1\\n                     /transl_table=11\\n                     /product=\"maturase K\"\\n                     /protein_id=\"ACC99456.1\"\\n                     /db_xref=\"GI:186972395\"\\n                     /translation=\"IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG\\n                     FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL\\n                     EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK\\n                     KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE\\n                     HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH\\n                     FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV\\n                     ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY\\n                     RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE\"\\nORIGIN      \\n        1 attttttacg aacctgtgga aatttttggt tatgacaata aatctagttt agtacttgtg\\n       61 aaacgtttaa ttactcgaat gtatcaacag aattttttga tttcttcggt taatgattct\\n      121 aaccaaaaag gattttgggg gcacaagcat tttttttctt ctcatttttc ttctcaaatg\\n      181 gtatcagaag gttttggagt cattctggaa attccattct cgtcgcaatt agtatcttct\\n      241 cttgaagaaa aaaaaatacc aaaatatcag aatttacgat ctattcattc aatatttccc\\n      301 tttttagaag acaaattttt acatttgaat tatgtgtcag atctactaat accccatccc\\n      361 atccatctgg aaatcttggt tcaaatcctt caatgccgga tcaaggatgt tccttctttg\\n      421 catttattgc gattgctttt ccacgaatat cataatttga atagtctcat tacttcaaag\\n      481 aaattcattt acgccttttc aaaaagaaag aaaagattcc tttggttact atataattct\\n      541 tatgtatatg aatgcgaata tctattccag tttcttcgta aacagtcttc ttatttacga\\n      601 tcaacatctt ctggagtctt tcttgagcga acacatttat atgtaaaaat agaacatctt\\n      661 ctagtagtgt gttgtaattc ttttcagagg atcctatgct ttctcaagga tcctttcatg\\n      721 cattatgttc gatatcaagg aaaagcaatt ctggcttcaa agggaactct tattctgatg\\n      781 aagaaatgga aatttcatct tgtgaatttt tggcaatctt attttcactt ttggtctcaa\\n      841 ccgtatagga ttcatataaa gcaattatcc aactattcct tctcttttct ggggtatttt\\n      901 tcaagtgtac tagaaaatca tttggtagta agaaatcaaa tgctagagaa ttcatttata\\n      961 ataaatcttc tgactaagaa attcgatacc atagccccag ttatttctct tattggatca\\n     1021 ttgtcgaaag ctcaattttg tactgtattg ggtcatccta ttagtaaacc gatctggacc\\n     1081 gatttctcgg attctgatat tcttgatcga ttttgccgga tatgtagaaa tctttgtcgt\\n     1141 tatcacagcg gatcctcaaa aaaacaggtt ttgtatcgta taaaatatat acttcgactt\\n     1201 tcgtgtgcta gaactttggc acggaaacat aaaagtacag tacgcacttt tatgcgaaga\\n     1261 ttaggttcgg gattattaga agaattcttt atggaagaag aa\\n//\\n\\n'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Cayla Mason\n",
    "# Section II\n",
    "#---------------------------------------------------------------------------------\n",
    "#BIOPYTHON EXERCISE - Biol568 BMI students\n",
    "#---------------------------------------------------------------------------------\n",
    "#\n",
    "# You will need to install Biopython and dependencies, or use a computer\n",
    "#   where they are installed.  https://biopython.org\n",
    "#\n",
    "# To write the functions, you can learn how to use Biopython using the\n",
    "#   Biopython Tutorial http://biopython.org/DIST/docs/tutorial/Tutorial.html\n",
    "#   I recommend trying out Biopython code on the command line first.\n",
    "\n",
    "# To use Biopython, you will need to import the Biopython libraries needed.\n",
    "# Follow the Tutorial. You will certainly need to read about how Biopython\n",
    "# stores sequence data.\n",
    "#\n",
    "#For example:\n",
    "#\n",
    "from Bio import SeqIO\n",
    "\n",
    "#\n",
    "#PART 1 - You will need a Fasta file. One with MULTIPLE SEQUENCE RECORDS.\n",
    "#   You can use the one in the tutorial (ls_orchid.fasta). \n",
    "#   See section 2.4.1  Simple FASTA parsing example\n",
    "\n",
    "#FUNCTION NAME: ParseFastaFile\n",
    "#PARAMETERS: 1 (File Name -a string of the name of a fasta sequence file)\n",
    "#PURPOSE: The function should:\n",
    "#           (1) Use Biopython to parse a fasta file with multiple sequences.\n",
    "#           (2) Make a dictionary: The keys will be the gi number \n",
    "#               and the values the sequences.\n",
    "#\n",
    "# The gi number is the first number of the title.\n",
    "# Example title sequence:\n",
    "# >gi|2765657|emb|Z78532.1|CCZ78532 C.californicum 5.8S rRNA gene and ITS1 and ITS2 DNA\n",
    "#\n",
    "# For this entry you will want '2765657' as the key.\n",
    "# This is the dictionary entry {'2765657':'CGTAACAAGGTTTCCGTAGGTGAACCTGCGGAAGGATCATTGTTGAGACAACAGAATATATGATCGAGTG...'}\n",
    "#\n",
    "#RETURN VALUES: The dictionary with ALL the entries.\n",
    "\n",
    "#== FUNCTION 1 ==\n",
    "def ParseFastaFile(f1):\n",
    "    sequences = {}\n",
    "    \n",
    "    for seq_record in SeqIO.parse(f1,\"fasta\"):\n",
    "        title = seq_record.id.split(\"|\")\n",
    "        key = title[1]\n",
    "        value = str(seq_record.seq)\n",
    "        sequences[key] = value\n",
    "        \n",
    "    return sequences \n",
    "\n",
    "#ParseFastaFile(\"ls_orchid.fasta.txt\")\n",
    "\n",
    "#---------------------------------------------------------------------------------\n",
    "#PART 2 - You will need a Genbank file. One with MULTIPLE SEQUENCE RECORDS.\n",
    "#   You can use the one in the tutorial (ls_orchid.gbk). \n",
    "#   See section 2.4.2  Simple GenBank parsing example\n",
    "#   and section 4.2.3  SeqRecord objects from GenBank files\n",
    "\n",
    "\n",
    "#FUNCTION NAME: GetGenbankInfo\n",
    "#PARAMETERS: 1 (File Name -a string of the name of a GenBank sequence file)\n",
    "#PURPOSE: The function should:\n",
    "#           (1) Use Biopython to parse the GenBank file with multiple sequences.\n",
    "#           (2) Make a dictionary -\n",
    "#                   (i)  The key will be the record.id\n",
    "#                   (ii) The value will be a LIST with two items:\n",
    "#                       1. the description\n",
    "#                       2. the sequence\n",
    "\n",
    "# For instance, from this record (only top part shown):\n",
    "\"\"\"\n",
    "LOCUS       NC_005816               9609 bp    DNA     circular BCT 21-JUL-2008\n",
    "DEFINITION  Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete\n",
    "            sequence.\n",
    "ACCESSION   NC_005816\n",
    "VERSION     NC_005816.1  GI:45478711\n",
    "PROJECT     GenomeProject:10638\n",
    "\"\"\"\n",
    "# The dictionary entry should be:\n",
    "# {'NC_005816.1': [''Yersinia pestis biovar Microtus str. 91001 plasmid pPCP1, complete sequence.', ''TGTAACGAACGGTGCAATAGTGATCCACACCCAACGCCTGAAATCAGATCCAGG...CTG']\n",
    "# \n",
    "#\n",
    "#RETURN VALUES: The dictionary with ALL the entries.\n",
    "#\n",
    "\n",
    "#== FUNCTION 2 ==\n",
    "def ParseGenbankFile(f1):\n",
    "    sequences = {}\n",
    "    \n",
    "    for seq_record in SeqIO.parse(f1, \"genbank\"):\n",
    "        key = seq_record.id\n",
    "        value = [seq_record.description , \n",
    "                 str(seq_record.seq)]\n",
    "        sequences[key] = value\n",
    "        \n",
    "    return sequences\n",
    "\n",
    "#print(ParseGenbankFile(\"ls_orchid.gbk.txt\"))\n",
    "\n",
    "#---------------------------------------------------------------------------------\n",
    "#PART 3 -   See sectons 3.8 Transcription AND 3.9  Translation\n",
    "#   Note: Assume the sequence is a CODING STRAND (don't reverse complement).\n",
    "\n",
    "#FUNCTION NAME: TranscribeTranslate\n",
    "#PARAMETERS: 1 (A list of sequence nmers)\n",
    "#PURPOSE: The function should:\n",
    "#           (1) Iterate through the LIST of DNA sequence nmers. You make them up.\n",
    "#           (2) Using Biopython, it should transcribe and then translate each sequence.\n",
    "#           (3) These amino acid sequences should be stored in a list.\n",
    "  \n",
    "#RETURN VALUES: A LIST of all the amino acid sequences (strings)\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.Alphabet import IUPAC\n",
    "\n",
    "#== FUNCTION 3 ==\n",
    "def TranscribeTranslate(listNmers):\n",
    "    proteins = []\n",
    "    \n",
    "    for nmer in listNmers:\n",
    "        codingDNA = Seq(nmer,IUPAC.unambiguous_dna)\n",
    "        #transcription\n",
    "        mRNA = codingDNA.transcribe()\n",
    "        #translation\n",
    "        proteins.append(str(mRNA.translate()))\n",
    "        \n",
    "    return proteins\n",
    "\n",
    "#nmers = ['GATGAGGAAC','ATGCCCACAGGACGACGA','GAGAGACCAAGCAGCCC']\n",
    "#print(TranscribeTranslate(nmers))\n",
    "\n",
    "\n",
    "#---------------------------------------------------------------------------------\n",
    "#PART 4 -  See section 9.6  EFetch: Downloading full records from Entrez\n",
    "#       This function will use a pubmed ID number to fetch a file from the internet\n",
    "#       and save it to a file on your computer, under the pubmed id name.\n",
    "\n",
    "#FUNCTION NAME: FetchGenbankFile\n",
    "#PARAMETERS: 1 (A string - Pubmed ID number)\n",
    "#PURPOSE: The function should:\n",
    "#           (1) Get the genbank file as text from the nucleotide database.\n",
    "#           (2) Write it to a file. (See below)\n",
    "#RETURN VALUES: 1 (True)\n",
    "\n",
    "from Bio import Entrez\n",
    "\n",
    "#== FUNCTION 4 ==\n",
    "def FetchGenbankFile(pubmedID):\n",
    "    Entrez.email = \"cayla.nicole.m@gmail.com\" #why is this necessary?\n",
    "    handle = Entrez.efetch(db=\"nucleotide\", id=pubmedID, rettype=\"gb\", retmode=\"text\")\n",
    "    try:\n",
    "        fileOut = open(pubmedID,'w')\n",
    "        fileOut.write(handle.read())\n",
    "        fileOut.close()\n",
    "        return 1\n",
    "    except:\n",
    "        return 0\n",
    "\n",
    "#x=FetchGenbankFile('186972394')\n",
    "\n",
    "\"\"\"\n",
    "If you were to call the FetchGenbankFile with the following id:\n",
    "\n",
    "x=FetchGenbankFile('186972394')\n",
    "\n",
    "You should get a file named 186972394.gbk.txt in your computer with the Genbank file below.\n",
    "\n",
    "From the tutorial:\n",
    "\n",
    ">>> from Bio import Entrez\n",
    ">>> Entrez.email = \"A.N.Other@example.com\"     # Always tell NCBI who you are\n",
    ">>> handle = Entrez.efetch(db=\"nucleotide\", id=\"186972394\", rettype=\"gb\", retmode=\"text\")\n",
    ">>> print handle.read()\n",
    "LOCUS       EU490707                1302 bp    DNA     linear   PLN 05-MAY-2008\n",
    "DEFINITION  Selenipedium aequinoctiale maturase K (matK) gene, partial cds;\n",
    "            chloroplast.\n",
    "ACCESSION   EU490707\n",
    "VERSION     EU490707.1  GI:186972394\n",
    "KEYWORDS    .\n",
    "SOURCE      chloroplast Selenipedium aequinoctiale\n",
    "  ORGANISM  Selenipedium aequinoctiale\n",
    "            Eukaryota; Viridiplantae; Streptophyta; Embryophyta; Tracheophyta;\n",
    "            Spermatophyta; Magnoliophyta; Liliopsida; Asparagales; Orchidaceae;\n",
    "            Cypripedioideae; Selenipedium.\n",
    "REFERENCE   1  (bases 1 to 1302)\n",
    "  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,\n",
    "            Endara,C.L., Williams,N.H. and Moore,M.J.\n",
    "  TITLE     Phylogenetic utility of ycf1 in orchids\n",
    "  JOURNAL   Unpublished\n",
    "REFERENCE   2  (bases 1 to 1302)\n",
    "  AUTHORS   Neubig,K.M., Whitten,W.M., Carlsward,B.S., Blanco,M.A.,\n",
    "            Endara,C.L., Williams,N.H. and Moore,M.J.\n",
    "  TITLE     Direct Submission\n",
    "  JOURNAL   Submitted (14-FEB-2008) Department of Botany, University of\n",
    "            Florida, 220 Bartram Hall, Gainesville, FL 32611-8526, USA\n",
    "FEATURES             Location/Qualifiers\n",
    "     source          1..1302\n",
    "                     /organism=\"Selenipedium aequinoctiale\"\n",
    "                     /organelle=\"plastid:chloroplast\"\n",
    "                     /mol_type=\"genomic DNA\"\n",
    "                     /specimen_voucher=\"FLAS:Blanco 2475\"\n",
    "                     /db_xref=\"taxon:256374\"\n",
    "     gene            <1..>1302\n",
    "                     /gene=\"matK\"\n",
    "     CDS             <1..>1302\n",
    "                     /gene=\"matK\"\n",
    "                     /codon_start=1\n",
    "                     /transl_table=11\n",
    "                     /product=\"maturase K\"\n",
    "                     /protein_id=\"ACC99456.1\"\n",
    "                     /db_xref=\"GI:186972395\"\n",
    "                     /translation=\"IFYEPVEIFGYDNKSSLVLVKRLITRMYQQNFLISSVNDSNQKG\n",
    "                     FWGHKHFFSSHFSSQMVSEGFGVILEIPFSSQLVSSLEEKKIPKYQNLRSIHSIFPFL\n",
    "                     EDKFLHLNYVSDLLIPHPIHLEILVQILQCRIKDVPSLHLLRLLFHEYHNLNSLITSK\n",
    "                     KFIYAFSKRKKRFLWLLYNSYVYECEYLFQFLRKQSSYLRSTSSGVFLERTHLYVKIE\n",
    "                     HLLVVCCNSFQRILCFLKDPFMHYVRYQGKAILASKGTLILMKKWKFHLVNFWQSYFH\n",
    "                     FWSQPYRIHIKQLSNYSFSFLGYFSSVLENHLVVRNQMLENSFIINLLTKKFDTIAPV\n",
    "                     ISLIGSLSKAQFCTVLGHPISKPIWTDFSDSDILDRFCRICRNLCRYHSGSSKKQVLY\n",
    "                     RIKYILRLSCARTLARKHKSTVRTFMRRLGSGLLEEFFMEEE\"\n",
    "ORIGIN      \n",
    "        1 attttttacg aacctgtgga aatttttggt tatgacaata aatctagttt agtacttgtg\n",
    "       61 aaacgtttaa ttactcgaat gtatcaacag aattttttga tttcttcggt taatgattct\n",
    "      121 aaccaaaaag gattttgggg gcacaagcat tttttttctt ctcatttttc ttctcaaatg\n",
    "      181 gtatcagaag gttttggagt cattctggaa attccattct cgtcgcaatt agtatcttct\n",
    "      241 cttgaagaaa aaaaaatacc aaaatatcag aatttacgat ctattcattc aatatttccc\n",
    "      301 tttttagaag acaaattttt acatttgaat tatgtgtcag atctactaat accccatccc\n",
    "      361 atccatctgg aaatcttggt tcaaatcctt caatgccgga tcaaggatgt tccttctttg\n",
    "      421 catttattgc gattgctttt ccacgaatat cataatttga atagtctcat tacttcaaag\n",
    "      481 aaattcattt acgccttttc aaaaagaaag aaaagattcc tttggttact atataattct\n",
    "      541 tatgtatatg aatgcgaata tctattccag tttcttcgta aacagtcttc ttatttacga\n",
    "      601 tcaacatctt ctggagtctt tcttgagcga acacatttat atgtaaaaat agaacatctt\n",
    "      661 ctagtagtgt gttgtaattc ttttcagagg atcctatgct ttctcaagga tcctttcatg\n",
    "      721 cattatgttc gatatcaagg aaaagcaatt ctggcttcaa agggaactct tattctgatg\n",
    "      781 aagaaatgga aatttcatct tgtgaatttt tggcaatctt attttcactt ttggtctcaa\n",
    "      841 ccgtatagga ttcatataaa gcaattatcc aactattcct tctcttttct ggggtatttt\n",
    "      901 tcaagtgtac tagaaaatca tttggtagta agaaatcaaa tgctagagaa ttcatttata\n",
    "      961 ataaatcttc tgactaagaa attcgatacc atagccccag ttatttctct tattggatca\n",
    "     1021 ttgtcgaaag ctcaattttg tactgtattg ggtcatccta ttagtaaacc gatctggacc\n",
    "     1081 gatttctcgg attctgatat tcttgatcga ttttgccgga tatgtagaaa tctttgtcgt\n",
    "     1141 tatcacagcg gatcctcaaa aaaacaggtt ttgtatcgta taaaatatat acttcgactt\n",
    "     1201 tcgtgtgcta gaactttggc acggaaacat aaaagtacag tacgcacttt tatgcgaaga\n",
    "     1261 ttaggttcgg gattattaga agaattcttt atggaagaag aa\n",
    "//\n",
    "\n",
    "\"\"\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
